### PR TITLE
refactor: Fix advanced compiler warnings

### PR DESCRIPTION
**Pull request by jamescowens**
This PR is an assortment of fixes in relatively fine-grained commits that clear up many of the warnings produced by compiling the Gridcoin code with advanced Clang compilers. Note that the first three commits are important, because the debug lock stack was not working correctly , so I ported the rest of the Bitcoin sync.h/cpp code (with a few necessary include tweaks). The last commit is also very important as #2270 broke RPC because ParseInt fails if lines have \r's.

The only thread safety work done here is the bare minimum required to stop Clang from complaining with AssertLockHeld.

Implementation of additional thread safety work, starting with the scraper, will be in separate PR's. In general the thread safety of Gridcoin has been good and wallet stability is very good, but Clang is quite exacting with the static thread analyzer, and this will require a significant amount of work to fully implement the GUARDED_BY and EXCLUSIVE_LOCKS_REQUIRED, etc.

### CODE REVIEW

File: src/policy/fees.h [hunk 1]
```diff
@@ -19,7 +20,19 @@ enum GetMinFee_mode
 inline CAmount GetBaseFee(const CTransaction& tx, enum GetMinFee_mode mode = GMF_BLOCK)
 {
     // Base fee is either MIN_TX_FEE or MIN_RELAY_TX_FEE
-    CAmount nBaseFee = (mode == GMF_RELAY) ? MIN_RELAY_TX_FEE : MIN_TX_FEE;
+
+    CAmount nBaseFee = 0;
+
+    // Writeen this way to silence the duplicate branch warning on advanced compilers while MIN_RELAY_TX_FEE
```
Comment (RoboticMind) [inline src/policy/fees.h]:
```suggestion
    // Written this way to silence the duplicate branch warning on advanced compilers while MIN_RELAY_TX_FEE
```

**Comment by jamescowens** [comment]
Ugh. The warnings on lock analysis related to AssertLockHeld with Clang I think will require the porting of the GUARDED_BY, etc. as implemented by Bitcoin now. See https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#negative.

**Comment by iFoggz** [comment]
:exploding_head:

**Comment by jamescowens** [comment]
@cyrossignol I am getting what I think are false-positives now on the thread-safety in the MacOS native CI Clang compilation output. Can you take a look and give me your opinion?

**Comment by jamescowens** [comment]
Temporarily this will show https://github.com/gridcoin-community/Gridcoin-Research/pull/2292/commits/d87e5d9ae54e9caf64b41a73791d4aec00016061, which is in #2305, until that one is merged and this one is rebased after the merge.

File: src/gridcoin/scraper/scraper_net.h [hunk 2]
```diff
@@ -127,41 +142,40 @@ class CScraperManifest
...
-        bool current =0;
-        bool last =0;
+        std::string project GUARDED_BY(cs_manifest);
+        std::string ETag GUARDED_BY(cs_manifest);
+        unsigned int LastModified GUARDED_BY(cs_manifest) = 0;
+        int part1 GUARDED_BY(cs_manifest) = -1;
+        unsigned partc GUARDED_BY(cs_manifest) = 0;
+        int GridcoinTeamID GUARDED_BY(cs_manifest) = -1;
+        bool current GUARDED_BY(cs_manifest) = 0;
+        bool last GUARDED_BY(cs_manifest) = 0;
```
Comment (cyrossignol) [inline src/gridcoin/scraper/scraper_net.h]: The `GUARDED_BY` attribute won't work here. The `dentry` class is related to `CScraperManifest` by name only&mdash;it is legal to instantiate these alone, so static analysis cannot assume that a `CScraperManifest` object will exist in scope to validate locks on `cs_manifest`. `CScraperManifest::projects` is the field we care about, and that one has the attribute already.

**Review by cyrossignol — COMMENTED** [review]
@jamescowens Are you referring to the bits like these?

<details>
<summary>Excerpt from native macOS build log <sup>[click to expand]</sup></summary>

```
     CXX      gridcoin/libgridcoin_util_a-tally.o
  gridcoin/quorum.cpp:1069:34: warning: reading variable 'project' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
    CXX      gridcoin/libgridcoin_util_a-tx_message.o
                      return entry.project == "VerifiedBeacons";
                                   ^
  gridcoin/quorum.cpp:1064:27: warning: reading variable 'projects' requires holding mutex 'manifest->cs_manifest' [-Wthread-safety-analysis]
                  manifest->projects.begin(),
                            ^
  gridcoin/quorum.cpp:1065:27: warning: reading variable 'projects' requires holding mutex 'manifest->cs_manifest' [-Wthread-safety-analysis]
                  manifest->projects.end(),
                            ^
  gridcoin/quorum.cpp:1079:69: warning: reading variable 'part1' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
              const size_t part_offset = verified_beacons_entry_iter->part1;
                                                                      ^
  gridcoin/quorum.cpp:1251:58: warning: passing variable 'project' by reference requires holding mutex 'cs_manifest' [-Wthread-safety-reference]
                  auto project_option = TallyProject(entry.project, scraper_id);
                                                           ^
  gridcoin/quorum.cpp:1256:46: warning: reading variable 'part1' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
                  if (!project_option || entry.part1 >= (int)manifest->vParts.size()) {
                                               ^
  gridcoin/quorum.cpp:1260:67: warning: reading variable 'part1' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
                  const uint256& part_hash = manifest->vParts[entry.part1]->hash;
                                                                    ^
  gridcoin/quorum.cpp:1266:48: warning: reading variable 'part1' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
                          manifest->vParts[entry.part1]->hash,
                                                 ^
  gridcoin/quorum.cpp:1268:31: warning: reading variable 'LastModified' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
                          entry.LastModified));
                                ^
  gridcoin/quorum.cpp:732:9: warning: unused member function 'ResolvedProject' [-Wunused-member-function]
          ResolvedProject() : m_combiner_mask(0)
          ^

[...]

  gridcoin/scraper/scraper.cpp:3837:49: warning: reading variable 'project' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
          else if (manifest->projects[iPartNum-1].project == "VerifiedBeacons")
                                                  ^
  gridcoin/scraper/scraper.cpp:3839:57: warning: reading variable 'project' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
              outputfile = manifest->projects[iPartNum-1].project + ".dat";
                                                          ^
  gridcoin/scraper/scraper.cpp:3843:57: warning: reading variable 'project' requires holding mutex 'cs_manifest' [-Wthread-safety-analysis]
              outputfile = manifest->projects[iPartNum-1].project + "-" + manifest->projects[iPartNum-1].ETag + ".csv.gz";
                                                          ^
  gridcoin/scraper/scraper.cpp:3843:104: warning: passing variable 'ETag' by reference requires holding mutex 'cs_manifest' [-Wthread-safety-reference]
              outputfile = manifest->projects[iPartNum-1].project + "-" + manifest->projects[iPartNum-1].ETag + ".csv.gz";
                                                                                                         ^
  gridcoin/scraper/scraper.cpp:4302:30: warning: writing variable 'project' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.project = "VerifiedBeacons";
                               ^
  gridcoin/scraper/scraper.cpp:4303:30: warning: writing variable 'LastModified' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.LastModified = ScraperVerifiedBeacons.timestamp;
                               ^
  gridcoin/scraper/scraper.cpp:4304:30: warning: writing variable 'current' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.current = true;
                               ^
  gridcoin/scraper/scraper.cpp:4307:30: warning: writing variable 'part1' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.part1 = iPartNum;
                               ^
  gridcoin/scraper/scraper.cpp:4308:30: warning: writing variable 'partc' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.partc = 0;
                               ^
  gridcoin/scraper/scraper.cpp:4309:30: warning: writing variable 'GridcoinTeamID' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.GridcoinTeamID = -1; //Not used anymore
                               ^
  gridcoin/scraper/scraper.cpp:4311:30: warning: writing variable 'last' requires holding mutex 'cs_manifest' exclusively [-Wthread-safety-analysis]
                  ProjectEntry.last = 1;
                               ^
```

</details>

Not exactly false positives, but not really meaningful warnings either... see review comment in scraper_net.h.

Sorry, haven't had time to review the rest... this PR is huge 🙂

File: src/gridcoin/scraper/scraper_net.h (hunk 2 above)
Comment (jamescowens) [reply src/gridcoin/scraper/scraper_net.h]: Aha! Thanks :)

**Comment by jamescowens** [comment]
I have tried to make the commits very organized so you can take them one by one! :)

**Comment by jamescowens** [comment]
This PR is big enough as it is. It only includes thread safety work sufficient to quiet down the Clang warnings for the AssertLockHeld. I am moving the more comprehensive thread safety work, starting with the scraper, to other PR's due to the large scope involved.

File: src/gridcoin/scraper/scraper_net.h (hunk 2 above)
Comment (jamescowens) [reply src/gridcoin/scraper/scraper_net.h]: Moving the scraper work to another PR

**Comment by jamescowens** [comment]
@div72 Please take a look at this if you have time. I would like to get it merged so I can straighten out the next tranch, which is #2315. Note that in https://github.com/gridcoin-community/Gridcoin-Research/pull/2292/commits/1229f27c5854bb592965a82a341e3c1aedbd969e, which is more or less a straight port of the newest Bitcoin sync.h/cpp code, I have tweaked it to allow continuing even if a potential deadlock is detected. This is necessary, because we have the cs_main cs_vNodes, etc. potential deadlock in our net code, which has been there for eternity, and doesn't actually result in an observed deadlock, but prevents running the wallet for any length of time to observe real potential deadlocks.

Note that it actually doesn't work, however. Sometimes after the first, second, or rarely third deadlock warning, the application segfaults on a destructor problem on the CLockLocation destructor. Looks like the address is pointing off into space. My guess is there is something wrong in the lock stack logic, but I can't see it. Maybe you can spot it.

This problem seriously impeded my scraper thread safety lock ordering work...

**Comment by div72** [comment]
@jamescowens The PR mostly looks good, there are a few formatting issues and possible SHARED_LOCKS_REQUIRES usages but they seem to be stemming from upstream so they aren't nit-worthy. One thing to note is that 1d745fe522c78bf2d32c8600ad9365d2080cb818 bumps our minimum required boost version to 1.60.

I'll check the deadlock warning issue tomorrow.

**Comment by div72** [comment]
@jamescowens Could you try removing the `lock_stack.pop_back();` in push_lock on L#151 of sync.cpp? To me it seems like that since the function was modified to continue execution, the pop_back there causes a double pop_back when the mutex gets unlocked.

**Comment by jamescowens** [comment]
Yep. That is a good idea. I should probably make it conditional on the flags since if the flags are set to true it should operate like the Bitcoin original

**Comment by jamescowens** [comment]
Yep. Looks like that worked.

**Comment by jamescowens** [comment]
@barton2526 Are we ok with a minimum boost of 1.60 or should I drop that commit?

**Comment by barton2526** [comment]
> 
> 
> @barton2526 Are we ok with a minimum boost of 1.60 or should I drop that commit?

Bitcoin set a minimum of 1.64 in July:
https://github.com/bitcoin/bitcoin/pull/22320

> The only system I can see that is affected here, (doesn't have new enough system packages) is Debian Oldstable

I am fine with it.

**Comment by caraka** [comment]
On the linux side, only CentOS 7 would be affected by a min boost 1.60. It has stopped receiving regular updates and only gets maintenance ones. Go for it.
Debian oldstable is not affected. Debian old oldstable, i.e. Stretch, is affected.

**Comment by jamescowens** [comment]
Very well.

**Comment by jamescowens** [comment]
@div72, I would like to get this one merged. You want to put in your comments/suggestions?

**Comment by div72** [comment]
nit: I'd squash functionally similar commits but this is also fine currently.

**Review by div72 — APPROVED** [review]

**Comment by jamescowens** [comment]
I was debating that too, but given any possible but unlikely regressions, this makes it easier to deal with....

**Pull request merged**